### PR TITLE
zml: emit `func.call` operations

### PR DESCRIPTION
**Pull request by gwenzek**
Restore the machinery to emit individual functions and calls to them.

this should unblock composites, and generates cleaner MLIR in general

* functions are cached based on the input shapes including tags.
* donations/tags are propagated through function calls

### CODE REVIEW

File: .github/workflows/commit.yaml
Comment (Corendos) [inline .github/workflows/commit.yaml:1]: This should probably be put in another commit/PR to have a proper history

File: zml/module.zig [hunk 1]
```diff
@@ -49,7 +49,7 @@ const Block = union(BlockKind) {
             .op_result => |parent_op| self.appendOperationRecursive(parent_op),
             .block_argument => |arg| {
                 // Hermetic blocks are not allowed to use arguments from other blocks.
-                std.debug.assert(self == .open or self.block().eql(arg.block()));
+                stdx.debug.assert(self == .open or self.block().eql(arg.block()), "Can't add {}  from {?x} block to {?x} block", .{ arg, arg.block()._inner.ptr, self.block()._inner.ptr });
```
Comment (Corendos) [inline zml/module.zig:52]:
```suggestion
                stdx.debug.assert(self == .open or self.block().eql(arg.block()), "Can't add {} from {?x} block to {?x} block", .{ arg, arg.block()._inner.ptr, self.block()._inner.ptr });
```

**Review by Corendos — APPROVED** [review]

File: .github/workflows/commit.yaml
Comment (gwenzek) [reply .github/workflows/commit.yaml:1]: done: https://github.com/zml/zml/pull/121

Comment (gwenzek) [reply .github/workflows/commit.yaml:1]: done https://github.com/zml/zml/pull/121
